apps/web/pages/api/auth/signup.ts

```diff
@@ -1,218 +1,50 @@
-          return res.status(422).json({
-            message: "Sign up from https://cal.com/signup to claim your premium username",
-          });
-        }
-      }
-
-      // Identify the org id in an org context signup, either the invited team is an org
-      // or has a parentId, otherwise parentId will be null, making orgId null
-      const orgId = teamMetadata?.isOrganization ? team.id : team.parentId;
-
-      const user = await prisma.user.upsert({
-        where: { email: userEmail },
-        update: {
-          username,
-          password: hashedPassword,
-          emailVerified: new Date(Date.now()),
-          identityProvider: IdentityProvider.CAL,
-          organizationId: orgId,
-        },
-        create: {
-          username,
-          email: userEmail,
-          password: hashedPassword,
-          identityProvider: IdentityProvider.CAL,
-          organizationId: orgId,
-        },
-      });
-
-      const membership = await prisma.membership.upsert({
-        where: {
-          userId_teamId: { userId: user.id, teamId: team.id },
-        },
-        update: {
-          accepted: true,
-        },
-        create: {
-          userId: user.id,
-          teamId: team.id,
-          accepted: true,
-          role: MembershipRole.MEMBER,
-        },
-      });
-      closeComUpsertTeamUser(team, user, membership.role);
-
-      // Accept any child team invites for orgs and create a membership for the org itself
-      if (team.parentId) {
-        // Create (when invite link is used) or Update (when regular email invitation is used) membership for the organization itself
-        await prisma.membership.upsert({
-          where: {
-            userId_teamId: { userId: user.id, teamId: team.parentId },
-          },
-          update: {
-            accepted: true,
-          },
-          create: {
-            userId: user.id,
-            teamId: team.parentId,
-            accepted: true,
-            role: MembershipRole.MEMBER,
-          },
-        });
-
-        // We do a membership update twice so we can join the ORG invite if the user is invited to a team witin a ORG
-        await prisma.membership.updateMany({
-          where: {
-            userId: user.id,
-            team: {
-              id: team.parentId,
-            },
-            accepted: false,
-          },
-          data: {
-            accepted: true,
-          },
-        });
-
-        // Join any other invites
-        await prisma.membership.updateMany({
-          where: {
-            userId: user.id,
-            team: {
-              parentId: team.parentId,
-            },
-            accepted: false,
-          },
-          data: {
-            accepted: true,
-          },
-        });
-      }
+export default async function handler(req: RequestWithUsernameStatus, res: NextApiResponse) {
+  // Use a try catch instead of returning res every time
```

Comment:
Suggestion to be considered at a later time: try/catch structures ideally are used to catch unintended errors within the application, not to control the flow of logic. By using the structure this way, we are adding complexity to the catch, which tends to complicate the code. We now intentionally throw an HttpError and we have to clutter the catch with a check of that type.

An alternative to this is to create well-structured validation classes. In said validation class, you would have a function like `validate()` which would run through all of the rules that need to be passed. If validation fails, you return a `ValidationResult` object with more information on why. From there, we update the response and return from the handler before executing other code.

To make it slightly cleaner on the handler side of things (a bit more complex on this dependency injection side), we could actually wrap this handler with what's called a "validation proxy". Basically, it's using the Proxy pattern for validation. You end up building a handler that is injected with a validation proxy and before the true handler is called, the proxy to perform validation is called first. They are not in the same file. This keeps things separate and we avoid polluting true handler code with ugly validation code.

Again, just a suggestion for something to look into for later to organize our validation in a different way.